src/inmanta/protocol/methods_v2.py

```diff
@@ -508,3 +508,51 @@ def resource_did_dependency_change(
     :raises BadRequest: When this endpoint in called while the resource with the given resource version is not
                         in the deploying state.
     """
+
+
+@typedmethod(path="/resource", operation="GET", arg_options=methods.ENV_OPTS, client_types=[ClientType.api], api_version=2)
+def resource_list(
+    tid: uuid.UUID,
+    limit: Optional[int] = None,
+    first_id: Optional[model.ResourceVersionIdStr] = None,
+    last_id: Optional[model.ResourceVersionIdStr] = None,
+    start: Optional[str] = None,
+    end: Optional[str] = None,
+    filter: Optional[Dict[str, List[str]]] = None,
+    sort: str = "resource_type.desc",
+) -> List[model.LatestReleasedResource]:
+    """
+    :param tid: The id of the environment this resource belongs to
+    :param limit: Limit the number of instances that are returned
+    :param first_id: The resource_version_id to use as a continuation token for paging, in combination with the 'start' value,
```

Comment:
Why not page purely on the resource_version_id? The resource_version_id is always unique.

Reply:
On a related note: how is the situation handled where this id no longer exists?

Reply:
For purely the paging, the `resource_version_id` could be enough, but to provide the metadata, both the `start` and `first_id` parameters are necessary. When counting the records that are not on the current page, we don't order or limit them, just count them if they are below the `start` and `first_id`.
If the id no longer exists, it was used for paging, and we are near the end of the table, then it's possible that that page will be smaller (or even empty). I understand this is not ideal, but I'm not sure what is the best solution, and if this should be handled on the server or client side.

Reply:
> but to provide the metadata, both the start and first_id parameters are necessary.

I still don't understand why that is.

Reply:
If I understand correctly, it's because when using an aggregate function, the ordering only comes into account after the calculation of the aggregate, so we can't sort the table first, and then use the ids to split the sorted result set into parts. To use the conditional aggregates for this, we need the `start` and `end` values, (and the ids to settle equality on the `order_by_column`).

Reply:
Could you point me to the query that has this problem?

Reply:
It's used in the `Resource._get_paging_item_count_query` method, the query starts on line 2792 (https://github.com/inmanta/inmanta-core/blob/28687dea90022e4fa84c4bc99c21ed1c5a36772f/src/inmanta/data/__init__.py#L2792)

Reply:
ah, I see, it is because we allow re-ordering